### Seurat - Guided Clustering Tutorial


<style type="text/css"> body, td { font-size: 14px; } code.r{ font-size: 12px; } pre { font-size: 12px } </style>

[Seurat - Guided Clustering Tutorial](https://http://satijalab.org/seurat/pbmc-tutorial.html)


```{r Prelims, include=F, echo=FALSE, results='hide', message=FALSE}

  FN <- "M2A-SeuratTutorial"

  suppressMessages(require(rmarkdown))
  suppressMessages(require(knitr))
  options(stringsAsFactors=F)

  # Shotcuts for knitting and redering while in R session
  kk <- function(n='') knit2html(paste("Scripts/t", n, sep=''), envir=globalenv(),
        output=paste('Scripts/',FN,".html", sep=''))
  bb <- function(n='') browseURL(paste('Scripts/',FN,".html", sep=''))

  # The usual shorcuts
  zz <- function(n='') source(paste("Scripts/t", n, sep=''))

  if(file.exists('/Users/francois/Dropbox/SingleCell/Jun2017/R/'))
  WRKDIR <- '/Users/francois/Dropbox/SingleCell/Jun2017/R/'  else
  WRKDIR <- '/mnt100/home/Dropbox/SingleCell/Jun2017/R'

  # do once
  #setwd(WRKDIR)

  # file rmarkdown file management options: cache, figures
  cache_DIR <- file.path(WRKDIR, 'Scripts', 'cache/M2A/')
  suppressMessages(dir.create(cache_DIR, recursive=T))
  opts_chunk$set(cache.path=cache_DIR)

  figure_DIR <- file.path(WRKDIR, 'Scripts', 'figure/M2A/')
  suppressMessages(dir.create(figure_DIR, recursive=T))
  opts_chunk$set(fig.path=figure_DIR)

  # Options from SingleCell RNAseq Workflow
  opts_chunk$set(error=FALSE, message=FALSE, warning=FALSE)
  opts_chunk$set(fig.width=11, fig.height=8)
  opts_chunk$set(dpi=300, dev="png", dev.args=list(pointsize=15))
  options(bitmapType="cairo", width=100)

```
<!-- ***************************************************** -->

```{r defFunctions, echo=F}
  # Place holder
```
<!-- ***************************************************** -->


```{r loadLibs, echo=F}
 #suppressMessages(require(Seurat))
 #suppressMessages(require(dplyr))
 #suppressMessages(require(Matrix))

 #suppressMessages(require(magrittr))

 #suppressMessages(require(data.table))
 #options(datatable.fread.datatable=F)

 #suppressMessages(require(RColorBrewer))

 #suppressMessages(require(edgeR))
 #suppressMessages(require(limma))

 #suppressMessages(require(Biobase))
 #suppressMessages(require(GenomicRanges))  ### PROBLEM on Mac OS!!!
                                           
 #suppressMessages(require(vcd))
 #suppressMessages(require(stringr))
 #suppressMessages(require(reshape2))
 #suppressMessages(require(tidyr))
 #suppressMessages(require(plyr))
 #suppressMessages(require(irlba))
 #suppressMessages(require(readr))
 #suppressMessages(require(scater))        ### PROBLEM on Mac OS!!!

 #suppressMessages(require(BiocStyle))

 #suppressMessages(require(R.utils))
 #suppressMessages(require(readxl))

 #suppressMessages(require(Rtsne))

 # ABLE TO INSTALL mvoutlier and destiny - Finally!
 #suppressMessages(require(mvoutlier))
 #suppressMessages(require(destiny))

 #suppressMessages(require(scran))

 # Setting single-core unless explicitly specified otherwise.
 suppressMessages(require(BiocParallel))
 register(SerialParam())

```
<!-- ***************************************************** -->


```{r setParam, echo=F}

 if(file.exists('/Users/francois/Dropbox/SingleCell/'))
 ROOT <- '/Users/francois/Dropbox/SingleCell/'  else
 ROOT <- '/mnt100/home/Dropbox/SingleCell/'

 EXT_DATA <- file.path(ROOT, 'Data', 'filtered_gene_bc_matrices', 'hg19')

 TABLE_DIR <- file.path(WRKDIR, 'Scripts', 'tables', 'M2A')
 suppressMessages(dir.create(TABLE_DIR,recursive=T))
 
```
<!-- ***************************************************** -->


### Setup the Seurat Object

In this tutorial, we will be analyzing the a dataset of Peripheral Blood Mononuclear Cells (PBMC) freely available from 10X Genomics. There are 2,700 single cells that were sequenced on the Illumina NextSeq 500. The raw data can be found [here](https://s3-us-west-2.amazonaws.com/10x.files/samples/cell/pbmc3k/pbmc3k_filtered_gene_bc_matrices.tar.gz).

We start by reading in the data. All features in Seurat have been configured to work with both regular and sparse matrices, but sparse matrices result in significant memory and speed savings for Drop-seq/inDrop/10x data.

```{r readData, cache=T}
 suppressMessages(require(Seurat))
 suppressMessages(require(dplyr))
 suppressMessages(require(Matrix))

 # Load the PBMC dataset
 pbmc.data <- Read10X(EXT_DATA)

 #Examine the memory savings between regular and sparse matrices
 dense.size <- object.size(as.matrix(pbmc.data))
 dense.size

 sparse.size <- object.size(pbmc.data)
 sparse.size

 dense.size/sparse.size

 # Initialize the Seurat object with the raw (non-normalized data)
 # Note that this is slightly different than the older Seurat workflow, 
 # where log-normalized values were passed in directly.
 # You can continue to pass in log-normalized values, just set do.logNormalize=F 
 # in the next step.
 pbmc <- new("seurat", raw.data = pbmc.data)

 # Keep all genes expressed in >= 3 cells, keep all cells with >= 200 genes
 # Perform log-normalization, first scaling each cell to a total of 1e4 molecules 
 # (as in Macosko et al. Cell 2015)
 pbmc <- Setup(pbmc, min.cells = 3, min.genes = 200, 
              do.logNormalize = T, total.expr = 1e4, project = "10X_PBMC")
```
<!-- ***************************************************** -->


### Basic QC and selecting cells for further analysis

While the setup function imposes a basic minimum gene-cutoff, you may want to filter out cells at this stage based on technical or biological parameters. Seurat allows you to easily explore QC metrics and filter cells based on any user-defined criteria. In the example below, we visualize gene and molecule counts, plot their relationship, and exclude cells with a clear outlier number of genes detected as potential **multiplets**. Of course this is not a guarenteed method to exclude **cell doublets**, but we include this as an example of filtering user-defined outlier cells. We also filter cells based on the percentage of mitochondrial genes present.

```{r basicQC, cache=T}

 # The number of genes and UMIs (nGene and nUMI) are automatically calculated for 
 # every object by Seurat.  For non-UMI data, nUMI represents the sum of the non-normalized 
 # values within a cell.  We calculate the percentage of mitochondrial genes here 
 # and store it in percent.mito using the AddMetaData. 
 # The % of UMI mapping to MT-genes is a common scRNA-seq QC metric.
 # NOTE: You must have the Matrix package loaded to calculate the percent.mito values. 
 mito.genes <- grep("^MT-", rownames(pbmc@data), value = T)
 percent.mito <- colSums(expm1(pbmc@data[mito.genes, ]))/colSums(expm1(pbmc@data))

 #n AddMetaData adds columns to object@data.info, and is a great place to stash QC stats
 pbmc <- AddMetaData(pbmc, percent.mito, "percent.mito")
 VlnPlot(pbmc, c("nGene", "nUMI", "percent.mito"), nCol = 3)

 # GenePlot is typically used to visualize gene-gene relationships, 
 # but can be used for anything calculated by the object, 
 # i.e. columns in object@data.info, PC scores etc.
 # Since there is a rare subset of cells with an outlier level of high 
 # mitochondrial percentage, and also low UMI content, we filter these as well
 par(mfrow = c(1, 2))
 GenePlot(pbmc, "nUMI", "percent.mito")
 GenePlot(pbmc, "nUMI", "nGene") 

 # We filter out cells that have unique gene counts over 2,500
 # Note that accept.high and accept.low can be used to define a 'gate', 
 # and can filter cells not only based on nGene but on anything in the object
 # (as in GenePlot above)
 pbmc <- SubsetData(pbmc, subset.name = "nGene", accept.high = 2500)
 pbmc <- SubsetData(pbmc, subset.name = "percent.mito", accept.high = 0.05)

 # Redo gene plots

 par(mfrow = c(1, 2))
 GenePlot(pbmc, "nUMI", "percent.mito")
 GenePlot(pbmc, "nUMI", "nGene") 
```

### Regress out unwanted sources of variation

Your single cell dataset likely contains ‘uninteresting’ sources of variation. This could include not only technical noise, but batch effects, or even biological sources of variation (cell cycle stage). As suggested in Buettner et al, NBT, 2015, regressing these signals out of the analysis can improve downstream dimensionality reduction and clustering. Seurat implements a basic version of this by constructing linear models to predict gene expression based on user-defined variables. Seurat stores the z-scored residuals of these models in the scale.data slot, and they are used for dimensionality reduction and clustering.

We typically regress out cell-cell variation in gene expression driven by batch (if applicable), cell alignment rate (as provided by Drop-seq tools for Drop-seq data), the number of detected molecules, and mitochondrial gene expression. For cycling cells, we can also learn a ‘cell-cycle’ score (as in Macosko et al) and Regress this out as well. In this simple example here for post-mitotic blood cells, we simply regress on the number of detected molecules per cell as well as the percentage mitochondrial gene content an example.

```{r RUV, cache=T}
 # Note that this overwrites pbmc@scale.data. Therefore, if you intend to use RegressOut, 
 # you can set do.scale=F and do.center=F in the original object to save some time.
 pbmc <- RegressOut(pbmc, latent.vars = c("nUMI", "percent.mito"))
```

### Detection of variable genes across the single cells

Seurat calculates highly variable genes and focuses on these for downstream analysis. **MeanVarPlot**(), which works by calculating the average expression and dispersion for each gene, placing these genes into bins, and then calculating a z-score for dispersion within each bin. This helps control for the relationship between variability and average expression. This function is unchanged from (Macosko et al.), but new methods for variable gene expression identification are coming soon. We suggest that users set these parameters to mark visual outliers on the dispersion plot, but the exact parameter settings may vary based on the data type, heterogeneity in the sample, and normalization strategy. The parameters here identify ~2,000 variable genes, and represent typical parameter settings for UMI data that is normalized to a total of 1e4 molecules.

```{r detectVG,cache=T}
 pbmc <- MeanVarPlot(pbmc ,fxn.x = expMean, fxn.y = logVarDivMean, 
         x.low.cutoff = 0.0125, x.high.cutoff = 3, y.cutoff = 0.5, do.contour = F)

 length(pbmc@var.genes)

```

### Perform linear dimensional reduction

Perform PCA on the scaled data. By default, the genes in object@var.genes are used as input, but can be defined using pc.genes. We have typically found that running dimensionality reduction on genes with high-dispersion can improve performance. However, with UMI data - particularly after using RegressOut, we often see that PCA returns similar (albeit slower) results when run on much larger subsets of genes, including the whole transcriptome.

```{r pca, cache=T}
 pbmc <- PCA(pbmc, pc.genes = pbmc@var.genes, do.print = TRUE, 
             pcs.print = 5, genes.print = 5)
```

ProjectPCA scores each gene in the dataset (including genes not included in the PCA) based on their correlation with the calculated components. Though we don’t use this further here, it can be used to identify markers that are strongly correlated with cellular heterogeneity, but may not have passed through variable gene selection. The results of the projected PCA can be explored by setting use.full=T in the functions below.

```{r ProjectPCA, cache=T}
 pbmc <- ProjectPCA(pbmc)
```

Seurat provides several useful ways of visualizing both cells and genes that define the PCA, including PrintPCA(), VizPCA(), PCAPlot(), and PCHeatmap()

```{r explorePCs, cache=T}
 PrintPCA(pbmc, pcs.print = 1:5, genes.print = 5, use.full = TRUE)

 VizPCA(pbmc, 1:2)

 PCAPlot(pbmc, 1, 2)

```

In particular PCHeatmap() allows for easy exploration of the primary sources of heterogeneity in a dataset, and can be useful when trying to decide which PCs to include for further downstream analyses. Both cells and genes are ordered according to their PCA scores. Setting cells.use to a number plots the ‘extreme’ cells on both ends of the spectrum, which dramatically speeds plotting for large datasets. Though clearly a supervised analysis, we find this to be a valuable tool for exploring correlated gene sets.


```{r PCHeatmap, cache=T}
 
 PCHeatmap(pbmc, pc.use = 1, cells.use = 100, do.balanced = TRUE)

 PCHeatmap(pbmc, pc.use = 1:12, cells.use = 500, do.balanced = TRUE,
           label.columns = FALSE, use.full = FALSE)

```

### Determine statistically significant principal components

To overcome the extensive technical noise in any single gene for scRNA-seq data, Seurat clusters cells based on their PCA scores, with each PC essentially representing a ‘metagene’ that combines information across a correlated gene set. Determining how many PCs to include downstream is therefore an important step.

In Macosko et al, we implemented a resampling test inspired by the jackStraw procedure. We randomly permute a subset of the data (1% by default) and rerun PCA, constructing a ‘null distribution’ of gene scores, and repeat this procedure. We identify ‘significant’ PCs as those who have a strong enrichment of low p-value genes.

```{r JackStraw, cache=T}
 # NOTE: This process can take a long time for big datasets, comment out for expediency.
 # More approximate techniques such as those implemented in PCElbowPlot() 
 # can be used to reduce computation time
 pbmc <- JackStraw(pbmc, num.replicate = 100, do.print = FALSE)
```

The JackStrawPlot() function provides a visualization tool for comparing the distribution of p-values for each PC with a uniform distribution (dashed line). ‘Significant’ PCs will show a strong enrichment of genes with low p-values (solid curve above the dashed line). In this case it appears that PCs 1-10 are significant.

```{r JackStrawPlot, cache=T}
JackStrawPlot(pbmc, PCs = 1:12)
```
A more ad hoc method for determining which PCs to use is to look at a plot of the standard deviations of the principle components and draw your cutoff where there is a clear elbow in the graph. This can be done with PCElbowPlot(). In this example, it looks like the elbow would fall around PC 9.

```{r PCElbowPlot, cache=T}
 PCElbowPlot(pbmc)
```

PC selection – identifying the true dimensionality of a dataset – is an important step for Seurat, but can be challenging/uncertain for the user. We therefore suggest these three approaches to consider. The first is more supervised, exploring PCs to determine relevant sources of heterogeneity, and could be used in conjunction with GSEA for example. The second implements a statistical test based on a random null model, but is time-consuming for large datasets, and may not return a clear PC cutoff. The third is a heuristic that is commonly used, and can be calculated instantly. In this example all three approaches yielded similar results, but we might have been justified in choosing anything between PC 7-10 as a cutoff. We followed the jackStraw here, admittedly buoyed by seeing the PCHeatmap returning interpretable signals (including canonical dendritic cell markers) throughout these PCs. Though the results are only subtly affected by small shifts in this cutoff (you can test below), we strongly suggest always explore the PCs they choose to include downstream.


### Cluster the cells

Seurat now includes an graph-based clustering approach compared to (Macosko et al.). Importantly, the *distance metric* which drives the clustering analysis (based on previously identified PCs) remains the same. However, our approach to partioning the cellular distance matrix into clusters has dramatically improved. Our approach was heavily inspired by recent manuscripts which applied graph-based clustering approaches to scRNA-seq data **[SNN-Cliq, Xu and Su, Bioinformatics, 2015]** and CyTOF data **[PhenoGraph, Levine et al., Cell, 2015]**. Briefly, these methods embed cells in a graph structure - for example a K-nearest neighbor (KNN) graph, with edges drawn between cells with similar gene expression patterns, and then attempt to partition this graph into highly interconnected ‘quasi-cliques’ or ‘communities’. As in PhenoGraph, we first construct a KNN graph based on the euclidean distance in PCA space, and refine the edge weights between any two cells based on the shared overlap in their local neighborhoods (Jaccard distance). To cluster the cells, we apply the smart local moving algorithm **[SLM, Blondel et al., Journal of Statistical Mechanics]**, to iteratively group cell groupings together with the goal of optimizing the standard modularity function.

The **FindClusters**() function implements the procedure, and contains a resolution parameter that sets the ‘granularity’ of the downstream clustering, with increased values leading to a greater number of clusters. We find that setting this parameter between 0.6-1.2 typically returns good results for single cell datasets of around 3K cells. Optimal resolution often increases for larger datasets. The clusters are saved in the object@ident slot.

#### HACK:
* To fix error:   cannot open file '/usr/local/lib/R/site-library/output39550.txt': No such file or directory
* sudo cp /usr/local/lib/R/site-library/edge_39550.txt /usr/local/lib/R/site-library/output39550.txt

#### STOP HERE - CAN'T GET FindClusters to work.

```{r FindClusters, cache=T}
 # save.SNN=T saves the SNN so that the  SLM algorithm can be rerun using 
 # the same graph, but with a different resolution value (see docs for full details)
 pbmc <- FindClusters(pbmc, pc.use = 1:10, resolution = 0.6, print.output = 0, save.SNN = T)
```

### Run Non-linear dimensional reduction (tSNE)

Seurat continues to use tSNE as a powerful tool to visualize and explore these datasets. While we no longer advise clustering directly on tSNE components, cells within the graph-based clusters determined above should co-localize on the tSNE plot. This is because the tSNE aims to place cells with similar local neighborhoods in high-dimensional space together in low-dimensional space. As input to the tSNE, we suggest using the same PCs as input to the clustering analysis, although computing the tSNE based on scaled gene expression is also supported using the genes.use argument.

```{r RunTSNE, cache=T}
 pbmc <- RunTSNE(pbmc, dims.use = 1:10, do.fast = T)

 #note that you can set do.label=T to help label individual clusters
 TSNEPlot(pbmc,do.label=T)

```

You can save the object at this point so that it can easily be loaded back in without having to rerun the computationally intensive steps performed above, or easily shared with collaborators.

```{r saveSeurat, cache=T}
  save(pbmc, file=file.path(WRKDIR, 'Data', 'pbmc'))
```

### Finding differentially expressed genes (cluster biomarkers)

Seurat can help you find markers that define clusters via differential expression. By default, it identifes positive and negative markers of a single cluster (specified in ident.1), compared to all other cells. **FindAllMarkers**() automates this process for all clusters, but you can also test groups of clusters vs. each other, or against all cells.

The *min.pct* argument requires a gene to be detected at a minimum percentage in either of the two groups of cells, and the *thresh.test* argument requires a gene to be differentially expressed (on average) by some amount between the two groups. You can set both of these to 0, but with a dramatic increase in time - since this will test a large number of genes that are unlikely to be highly discriminatory. As another option to speed up these computations, *max.cells.per.ident* can be set. This will downsample each identity class to have no more cells than whatever this is set to. While there is generally going to be a loss in power, the speed increases can be signficiant and the most highly differentially expressed genes will likely still rise to the top.

```{r fiundClusterBioMarkers, cahce=T}
 # find all markers of cluster 1
 cluster1.markers <- FindMarkers(pbmc, ident.1 = 1, min.pct = 0.25)
 print(head(cluster1.markers, 5))

 # find all markers distinguishing cluster 5 from clusters 0 and 3
 cluster5.markers <- FindMarkers(pbmc, 5, c(0,3), min.pct = 0.25)
 print(head(cluster5.markers, 5))

 # find markers for every cluster compared to all remaining cells, \
 # report only the positive ones
 pbmc.markers <- FindAllMarkers(pbmc, only.pos = TRUE, min.pct = 0.25, thresh.use = 0.25)
 pbmc.markers %>% group_by(cluster) %>% top_n(2, avg_diff)

```

Seurat has four tests for differential expression which can be set with the test.use parameter: ROC test (“roc”), t-test (“t”), LRT test based on zero-inflated data (“bimod”, default), LRT test based on tobit-censoring models (“tobit”) The ROC test returns the ‘classification power’ for any individual marker (ranging from 0 - random, to 1 - perfect).

```{r clusterMarkers, cache=T}
 cluster1.markers <- FindMarkers(pbmc, ident.1 = 0, thresh.use = 0.25, test.use = "roc",only.pos = T)
```
There are several tools for visualizing marker expression. VlnPlot() generates a violin plot which shows the probability density at different expression levels of the gene for each cluster. As seen below, good marker genes will show strongly in a single cluster. It can also be useful to look at gene/gene correlation with GenePlot() which returns a plot similar to a ‘FACS’ plot with cells colored by cluster identity. Also, the FeaturePlot() function is useful for viewing the expression of the gene in the context of all the cells and helps validate the specificity of the marker or the quality of the clustering.


```{r vizmarkers, cache=T}
 VlnPlot(pbmc, c("MS4A1","CD79A"))

 #you can plot raw UMI counts as well
 VlnPlot(pbmc, c("NKG7","PF4"),use.raw = T,y.log = T)

 FeaturePlot(pbmc, c("MS4A1", "GNLY","CD3E","CD14","FCER1A","FCGR3A", "LYZ", "PPBP", "CD8A"),cols.use = c("grey","blue"))

```

Heatmaps can also be a good way to examine heterogeneity within/between clusters. The DoHeatmap() function will generate a heatmap for given cells and genes. In this case, we are plotting the top 20 markers (or all markers if less than 20) for each cluster.


```{r heatMapBioMark, cache=T}

 pbmc.markers %>% group_by(cluster) %>% top_n(10, avg_diff) -> top10
 # setting slim.col.label to TRUE will print just the cluster IDS instead of every cell name
 DoHeatmap(pbmc, genes.use = top10$gene, order.by.ident = TRUE, 
         slim.col.label = TRUE, remove.key = TRUE)

```

### Assigning cell type identity to clusters

```{r, cache=T}
current.cluster.ids <- c(0, 1, 2, 3, 4, 5, 6, 7)
new.cluster.ids <- c("CD4 T cells", "CD14+ Monocytes", "B cells", "CD8 T cells", "FCGR3A+ Monocytes", "NK cells", "Dendritic cells", "Megakaryocytes")
pbmc@ident <- plyr::mapvalues(pbmc@ident, from = current.cluster.ids, to = new.cluster.ids)
TSNEPlot(pbmc, do.label = T, pt.size = 0.5)
```

### Further subdivisions within cell types

If you perturb some of our parameter choices above (for example, setting resolution=0.8 or changing the number of PCs), you might see the CD4 T cells subdivide into two groups. You can explore this subdivision to find markers separating the two T cell subsets. However, before reclustering (which will overwrite object@ident), we can stash our renamed identities to be easily recovered later.



```{r, cache=T}

#First lets stash our identities for later
pbmc <- StashIdent(pbmc, save.name = "ClusterNames_0.6")

#Note that if you set save.snn=T above, you don't need to recalculate the SNN, and can simply put : pbmc=FindClusters(pbmc,resolution = 0.8)
pbmc <- FindClusters(pbmc, pc.use = 1:10, resolution = 0.8, print.output = F)

#Demonstration of how to plot two tSNE plots side by side, and how to color points based on different criteria
plot1 <- TSNEPlot(pbmc, do.return = T, no.legend = TRUE, do.label = T)
plot2 <- TSNEPlot(pbmc, do.return = T, group.by = "ClusterNames_0.6", no.legend = TRUE, do.label = T)
MultiPlotList(list(plot1, plot2), cols = 2)

#Find discriminating markers
tcell.markers <- FindMarkers(pbmc, 0, 1)

#Most of the markers tend to be expressed in C1 (i.e. S100A4). However, we can see that CCR7 is upregulated in C0, strongly indicating that we can differentiate memory from naive CD4 cells.
#cols.use demarcates the color palette from low to high expression
FeaturePlot(pbmc, c("S100A4", "CCR7"), cols.use = c("green", "blue"))

```
 
The memory/naive split is bit weak, and we would probably benefit from looking at more cells to see if this becomes more convincing (stay tuned!). In the meantime, we can restore our old cluster identities for downstream processing.


```{r, cache=T}

pbmc <- SetAllIdent(pbmc, id = "ClusterNames_0.6")


```


### Parameter settings:
  * WRKDIR = `r WRKDIR`
  * FN = `r FN`
  * Scripts = Scripts
  * EXT_DATA = `r EXT_DATA`
  * TABLE_DIR = `r TABLE_DIR` 
  * RUN DATE = `r date()`


```{r, echo=F}
 sessionInfo()
### STOP HERE
  knit_exit()
```


<!-- To run

nohup sudo Rscript -e "knitr::knit2html('M2A-SeuratTutorial.Rmd')" > M2A-SeuratTutorial.log  &

-->

